FixMath.h

```diff
@@ -0,0 +1,1047 @@
+    - all operations between fixed point number is safe (it won't overflow) and preserve the precision. In particular:
+    - only addition, subtraction and multiplication are implemented
+    - any operation between a signed and an unsigned leads to a signed number
+    - resulting numbers will be casted to a type big enough to store the expected values. It follows that it is worth starting with types that are as small as possible to hold the initial value.
+    - all operations between a fixed point number and a native type (int, float, uint) are *not* safe. If the resulting value cannot be represented in the fixed point type it will overflow. Only addition, subtraction, multiplication and right/left shift are implemented.
+    - safe right/left shifts, which return the correct value in the correct type are implemented as .sR<shift>() and .sL<shift>() respectively, shift being the shifting amount. These shifts are basically on
+
+    More specifically on the returned types of the operations between fixed point math types:
+    - Additions:
+    - UFixMath<NI,NF> + UFixMath<_NI,_NF> returns UFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - SFixMath<NI,NF> + SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> + SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> + anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> + anything_else returns SFixMath<NI,NF>
+    - Subtractions:
+    - UFixMath<NI,NF> - UFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI),MAX(NF,_NF)>
+    - SFixMath<NI,NF> - SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - SFixMath<NI,NF> - UFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> - anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> - anything_else returns SFixMath<NI,NF>
+    - (-)SFixMath<NI,NF> return SFixMath<NI,NF>
+    - (-)UFixMath<NI,NF> return SFixMath<NI,NF>
+    - Multiplications:
+    - UFixMath<NI,NF> * UFixMath<_NI,_NF> returns UFixMath<NI+_NI,NF+_NF>
+    - UFixMath<NI,NF> * SFixMath<_NI,_NF> returns SFixMath<NI+_NI,NF+_NF>
+    - SFixMath<NI,NF> * SFixMath<_NI,_NF> returns SFixMath<NI+_NI,NF+_NF>
+    - UFixMath<NI,NF> * anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> * anything_else returns SFixMath<NI,NF>
+    - Shifts:
+    - UFixMath<NI,NF> .sR<NS> returns UFixMath<NI-NS,NF+NS>
+    - UFixMath<NI,NF> .sL<NS> returns UFixMath<NI+NS,NF-NS>
+    - same for SFixMath.
+      
+   
+*/
+
+
+
+
+#ifndef FIXMATH2_H_
+#define FIXMATH2_H_
+
+#include "IntegerType.h"
+
+#define MOZZI_SHIFTR(x,bits) (bits > 0 ? (x >> (bits)) : (x << (-bits))) // shift right for positive shift numbers, and left for negative ones.
+#define MAX(N1,N2) (N1 > N2 ? N1 : N2)
+#define UBITSTOBYTES(N) (((N-1)>>3)+1)
+#define SBITSTOBYTES(N) (((N)>>3)+1)
+#define ONESBITMASK(N) ((1ULL<<(N)) - 1)
```

Comment:
Question for @tfry-git : will this be optimized by the compiler in terms of type (even with the `ULL`?). I mean, `ONESBITMASK(8)` should be: 0b11111111 ideally, but is the `ULL` leading to 0b000000000000000000011111111?
Seems okay on gobolt, but prefer to be sure.

Reply:
> Question for @tfry-git : will this be optimized by the compiler in terms of type (even with the `ULL`?). I mean, `ONESBITMASK(8)` should be: 0b11111111 ideally, but is the `ULL` leading to 0b000000000000000000011111111? Seems okay on gobolt, but prefer to be sure.

That's always difficult to tell, as optimization is always a "may", and never a "must". But I do think it's a fairly safe bet in this case: As long as N is known at compile time, so is the result, and the compiler will surely be smart enough to figure out what data type it fits into. Now, in the case of integer division, that also clearly limits the range of the result of the division. And then, shortly after, the result is cast to internal_type. Of course things would stand quite different, if other operations would be involved.

---

Suggested alterative
```
[inside UFixMath itself:]
private:
  static [constexpr if possible] internal_type onesbitmask() { return ((1ULL << (NI+NF)) - 1); }
```

---

> Another one: should everything in there be inline to avoid clashes with user cpp files in your opinion?

Yes, indeed. More specifically, it's the one-defintion-rule, again, when more than one .cpp file is involved. That should "only" concern the operators, AFAICS, though: Definitions inside a class-declaration are automatically inline.

Reply:
> Edit: Or even, avoiding your initial question:
`static [constexpr if possible] internal_type onesbitmask() { return (internal_type(1) << (NI+NF)) - 1); }`

Unfortunately, I think this provokes an overflow if NI+NF=8 or 16 or 32, as the idea is to have all bits set to one, so we take the next value and substract one. A for loop might work in the internal type but I need to see if `constexpr` allows it.

Otherwise, from tests with gobolt, it seems that the compiler do optimize it away.

Reply:
> A for loop might work in the internal type but I need to see if constexpr allows it.

It does… After C++14… So no luck there. Your initial proposition do work, but I guess this is more or like the macro in effect.

Reply:
`static constexpr` work. Two of them are needed because the accurate division needs a bigger one.

From compiler statistics, it seems to reduce quite a bit the prog mem compared to macros, but I haven't checked which one do perform the best in terms of speed. Committed that for now nevertheless and will try to do some perf check soon. I'll squash a bit my commits at the end, this completely exploded…

Reply:
Final word, `constexpr` version is *way* faster… Don't fully understand why but this is striking: 3240us for a thousand `invFast()` original, 2040us for the one using `constexpr`, this is massive!

Will switch it all to `constexpr` and retest fully the computations.